src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -3489,39 +3512,30 @@ public void reconnect(final boolean shouldRequestVersion) throws IOException, SQ
 				throw new IOException();
 			}
 
-			try
-			{
-				clientHandshake(user, pwd, database, shouldRequestVersion);
-				clientHandshakeFinalize();
-
-				return;
-			}
-			catch (final Exception handshakeException)
-			{
-				try
-				{
-					LOGGER.log(Level.INFO, "[nio] XGConnection in.close");
-					closeTransport();
-				}
-				catch (final IOException f)
-				{
-				}
+			return clientHandshakeAsync(user, pwd, database, shouldRequestVersion)
+					.thenCompose(successOrEx -> {
+						
+						if(successOrEx.isException()) {
+							try
+							{
+								LOGGER.log(Level.INFO, "[nio] XGConnection in.close");
+								closeTransport();
```

Comment:
This synchronous call should be issued in a separate thread than the callback via `thenComposeAsync(...)` OR by using the helper method `runAsyncSafe_FIXME(...)`.
```suggestion
								return runAsyncSafe_FIXME(this::closeTransport())
								    .thenApply(constant(successOrEx));
```